browser/src/control/jsdialog/Widget.ColorPicker.ts

```diff
@@ -103,49 +103,266 @@ function createColor(
+						const [row] = getRowColumn(el);
+						return row === currentRow;
+					}).length - 1; // Move to the last column in the same row
+			}
+		}
+	} else if (axis === 'vertical') {
+		if (direction === 'next') {
+			targetRow++;
+		} else {
+			targetRow--;
 		}
+	}
 
-		const recentRow = palette[palette.length - 1];
-		if (recentRow.indexOf(colorItem) !== -1)
-			recentRow.splice(recentRow.indexOf(colorItem), 1);
-		recentRow.unshift(colorItem);
-		window.prefs.set('recentColor', JSON.stringify(recentRow));
+	// Find the target element based on the calculated row and column
+	const targetElement = focusableElements.find((el) => {
+		const [row, column] = getRowColumn(el);
+		return row === targetRow && column === targetColumn;
 	});
 
-	return color;
+	if (!targetElement && axis === 'vertical') {
+		if (direction === 'next') {
+			// Start from the next sibling of the parent container
+			const nextFocusableElement = findFocusableElement(
+				nextElement as HTMLElement,
+				'next',
+				isFocusable,
+			);
+			if (nextFocusableElement) {
+				nextFocusableElement.focus();
+			}
+		} else if (direction === 'previous') {
+			// Start from the previous sibling of the parent container
+			const previousFocusableElement = findFocusableElement(
+				nextElement as HTMLElement,
+				'previous',
+				isFocusable,
+			);
+			if (previousFocusableElement) {
+				previousFocusableElement.focus();
+			}
+		}
+	}
+
+	if (targetElement) {
+		targetElement.focus();
+	}
+}
+
+function findFocusableElement(
+	element: HTMLElement,
+	direction: 'next' | 'previous',
+	isFocusable: (el: HTMLElement) => boolean,
+): HTMLElement | null {
+	if (!element) {
+		const mainColorContainer = document.querySelector('.ui-color-picker');
+		const focusableElements = Array.from(
+			mainColorContainer.querySelectorAll('*'),
+		);
+		const firstFocusableElement =
+			direction === 'next'
+				? (focusableElements.find(isFocusable) as HTMLElement)
+				: (focusableElements.reverse().find(isFocusable) as HTMLElement);
+		if (firstFocusableElement) {
+			return firstFocusableElement;
+		}
+	}
+
+	// Check the current element if it is focusable
+	if (isFocusable(element)) return element;
+
+	// Check if sibling is focusable or contains focusable elements
+	const focusableInSibling = findFocusableWithin(
+		element as HTMLElement,
+		direction,
+	);
+	if (focusableInSibling) return focusableInSibling;
+
+	// Depending on the direction, find the next or previous sibling
+	const sibling: Element =
+		direction === 'next'
+			? element.nextElementSibling
+			: element.previousElementSibling;
+
+	if (sibling) {
+		// Recursively check the next or previous sibling of the current sibling
+		return findFocusableElement(sibling as HTMLElement, direction, isFocusable);
+	}
+
+	return null;
+}
+
+// Helper function to find the first focusable element within an element
+function findFocusableWithin(
+	element: HTMLElement,
+	direction: string,
+): HTMLElement | null {
```

Comment:
we already have something similar: https://github.com/CollaboraOnline/online/blob/0e109494b536a52e11c04d8a7017c285a9bf004f/browser/src/control/jsdialog/Util.FocusCycle.js#L18

can we reuse old one?

Reply:
all "focus" related things could go to `Util.FocusCycle.js` and new helper will be only using them and will be for "keyboard handling" part